ISISComputingGroup/EPICS-ioc
# Ticket2173 HRPD Sample Changer

### Description of work

Added support for the HRPD sample changer, including IOC, OPI and emulator as per https://github.com/ISISComputingGroup/IBEX/issues/2173.

EPICS-Top: https://github.com/ISISComputingGroup/EPICS/pull/80
Emulator: https://github.com/ISISComputingGroup/EPICS-DeviceEmulator/pull/12
Support: https://github.com/ISISComputingGroup/EPICS-rotating_sample_changer
GUI: https://github.com/ISISComputingGroup/ibex_gui/pull/480

### To test

The basics:
* Start the IOC in devsim 
* Start the emulator on the port the IOC is looking on
* Confirm that the IOC connects w/o error, all PVs appear to connect
* Create a device screen for the sample changer
* Confirm that the screen is all connected ok and that you can initialise then move the device

Testing against the VI:
* Connect the HRPD sample changer VI to the emulator (https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Emulating-Devices#connecting-an-emulator-to-labview)
* Confirm that the VI behaves in the same way as the OPI+IOC

Some slight niggles:
* The way the emulator handles errors is a bit dodgy, with no access to the device I just put in as much as I needed to get going
* The VI expects something slightly different from the documented API, I went with what the VI expects
* Some functions of the sample changer are HRPD specific, as default these aren't loaded but they are if IS_HRPD=''. I didn't expose this macro to the GUI as it's not something the user will need to change

---

#### Code Review


### Functional Tests


### Final steps
